File: internal/pkg/dto/response/conversation_memory.go
package response

import (
	"agent-desk/internal/pkg/reception"
	"time"
)

type MemorySource struct {
	ID             int64     `json:"id"`
	ConversationID int64     `json:"conversationId"`
	SenderType     string    `json:"senderType"`
	Content        string    `json:"content"`
	CreatedAt      time.Time `json:"createdAt"`
}

type MemoryEntry struct {
	ID             int64          `json:"id"`
	ConversationID int64          `json:"conversationId"`
	Kind           string         `json:"kind"`
	FieldKey       string         `json:"fieldKey"`
	Topic          string         `json:"topic"`
	Label          string         `json:"label"`
	Value          string         `json:"value"`
	Confirmed      bool           `json:"confirmed"`
	Revision       int64          `json:"revision"`
	UpdatedAt      time.Time      `json:"updatedAt"`
	Sources        []MemorySource `json:"sources"`
}

type ConversationMemory struct {
	ReceptionPolicy    reception.Policy `json:"receptionPolicy"`
	Status             string           `json:"status"`
	ErrorCode          string           `json:"errorCode"`
	AttemptCount       int              `json:"attemptCount"`
	MaxAttempts        int              `json:"maxAttempts"`
	NextRetryAt        *time.Time       `json:"nextRetryAt,omitempty"`
	ProcessedMessageID int64            `json:"processedMessageId"`
	UpdatedAt          *time.Time       `json:"updatedAt,omitempty"`
	HandoffReason      string           `json:"handoffReason"`
	Entries            []MemoryEntry    `json:"entries"`
	Shared             []MemoryEntry    `json:"shared"`
	Dossier            []MemoryEntry    `json:"dossier"`
}

File: web/lib/api/conversation-memory.ts
import { request } from "@/lib/api/client";
import { MemoryKind, type MemoryStatus } from "@/lib/generated/enums";
import type { ReceptionPolicy } from "@/lib/reception";

export type MemoryEntry = {
  id: number;
  conversationId: number;
  kind: MemoryKind;
  fieldKey?: string;
  topic: string;
  label: string;
  value: string;
  confirmed: boolean;
  revision: number;
  updatedAt: string;
  sources: { id: number; conversationId: number; senderType: string; content: string; createdAt: string }[];
};

export type ConversationMemory = {
  receptionPolicy?: ReceptionPolicy;
  status: MemoryStatus;
  errorCode: string;
  attemptCount: number;
  maxAttempts: number;
  nextRetryAt?: string;
  processedMessageId: number;
  updatedAt?: string;
  handoffReason: string;
  entries: MemoryEntry[];
  shared: MemoryEntry[];
  dossier?: MemoryEntry[];
};

export function fetchConversationMemory(id: number) {
  return request<ConversationMemory>(`/api/dashboard/conversation/${id}/memory`);
}

export function refreshConversationMemory(id: number) {
  return request<void>(`/api/dashboard/conversation/${id}/memory/refresh`, { method: "POST" });
}

export function updateConversationMemory(conversationId: number, entry: MemoryEntry, value: string, remove = false) {
  const action = entry.kind === MemoryKind.CustomerProfile ? "profile/update" : "update";
  return request<void>(`/api/dashboard/conversation/${conversationId}/memory/${action}`, {
    method: "POST", body: JSON.stringify({ id: entry.id, revision: entry.revision, value, delete: remove }),
  });
}
